### ISSUE TITLE
Support For CPU Atomics

### ISSUE THREAD

**Issue by insertinterestingnamehere** [numba/numba#2988]
## Feature request

It would be great if there were a way to use atomic operations from within numba. Currently atomic operations are only supported when generating cuda or HSA code.

I'm currently experimenting with using numba to write callbacks that are provided to a C++ concurrency library. In this context, atomics are useful in that they can reduce the amount of synchronization required (or to write the locks that are actually being used). You can technically use atomic operations now if you hide them inside a function pointer exported from C++, but the performance impacts of doing so aren't clear to me. At first glance, preventing inlining seems like a high price to pay for every atomic operation.

With regards to what would be the most useful, weak memory consistency is by far the most common thing that shows up in the applications I'm looking at right now, but what's needed is application specific in general. Other consistencies could potentially be useful to me for being able to acquire/release existing locks without having to route everything through an externally compiled function, but that's more of a secondary concern at the moment.

**Comment by s22chan** [comment]
FYI I was just looking at this and you can achieve it reasonably with https://gist.github.com/sklam/40f25167351832fe55b64232785d036d (renamed to `numba.extending.intrinsics`), but it requires a bit of llvm work.

**Comment by insertinterestingnamehere** [comment]
Neat example! I hadn't seen that before.

**Comment by mdekstrand** [comment]
To document a use case for this feature: the [HogWild! algorithm](http://papers.nips.cc/paper/4390-hogwild-a-lock-free-approach-to-parallelizing-stochastic-gradient-descent) requires atomic floating-point addition. While that may be achievable in Numba's CUDA support, we need atomic intrinsics (ideally applicable to NumPy arrays, an intrinsic atomic `arr[i] += x` operation) in order to make such implementations that work without a GPU.

**Comment by vlad17** [comment]
Just wanted to add to this, I’ve felt a need for this when I wanted to:

* generate large histograms in parallel (to large to give each thread a copy)
* perform GraphBLAS-like parallel graph computations

**Comment by insertinterestingnamehere** [comment]
Yep, my use-cases involve parallel graph analytics as well.

**Comment by EdwardRaff** [comment]
Adding another request for CPU atomics. I'm not sure how multi-dimensional cases are handled, but abstractly if there was an `atomic_add(arr, indx, val)` `atomic_compare_and_swap(arr, index, expected, new)` and an `atomig_get(arr, index)` that would be enough for me to work around a lack of other operations and get stuff working. This would let me get some [n-gram based](https://arxiv.org/abs/1908.00200) parallel computing working and be enough for a lot of common workloads like Hogwild.

**Comment by slayoo** [comment]
voting up! Something like CUDA's atomicAdd, atomicMin, etc would be helpful.
For the OMP hackend, the ``#pragma omp atomic`` should be of some help?

**Comment by sklam** [comment]
I attempted to write an atomic add using `@intrinsic` about it is held up by https://github.com/numba/numba/issues/6920.

**Comment by insertinterestingnamehere** [comment]
CC @arthurp. Didn't you hack something like this together?

**Comment by arthurp** [comment]
Yes. I don't love the API, but it does work: https://github.com/KatanaGraph/katana/blob/master/python/katana/numba_support/numpy_atomic.py (The project is BSD licensed)

Ideally there would be some way to mark a normal array access as needing atomic code gen. 

My hack directly generates LLVM instructions for the atomics. It should also implement CAS loops for operations that are not available as single LLVM instructions, but it doesn't yet.

*My code may be memory management unsafe:* The system it's used with only uses very limited compiled functions and they are always called synchronously from within a python call, so we don't need to reference count anything.

I hope this code is useful.

**Comment by EdwardRaff** [comment]
@arthurp that looks interesting. Could you elaborate further on when it may not work? If I pass a numpy array in, the original code has the pointer/ref count, I'm not sure I'm following your explanation of when/where the risk occurs? Presumably that would be fine with multiple threads via the prange function 

Also, which dtypes does it support?

**Comment by arthurp** [comment]
>  Could you elaborate further on when it may not work? If I pass a numpy array in, the original code has the pointer/ref count, I'm not sure I'm following your explanation of when/where the risk occurs? Presumably that would be fine with multiple threads via the prange function

You know. I'm thinking back on this and I think this atomic code should be safe. I was confusing this hack with other hacks I did around the same time. I think the main problem with this code is probably mostly that I suspect it can generate invalid LLVM code if some dtypes are passed in, since I don't check which LLVM instructions really exist (the atomic instructions only exist at specific types). I also never created the CAS loop fallback.

> Also, which dtypes does it support?

Integers up to 64 bits (probably? it might just be 32 and 64-bits), and floats of 32 and 64 bits. I think. Those are definitely the only ones I've tested.

**Comment by yojeep** [comment]
voting up!

**Comment by slayoo** [comment]
CC @arthurp, @Sfonxu 

For the record, the `numpy_atomic.py` file referenced above has moved to:
https://github.com/KatanaGraph/katana/blob/master/python/katana/native_interfacing/numpy_atomic.py

I confirm it used to work great ([used in the PySDM project](https://github.com/open-atmos/PySDM/blob/main/PySDM/backends/impl_numba/atomic_operations.py)), however this workaround does not seem to work on ARM CPUs. Any ideas on how to achieve atomic operations in multi-threaded Numba JIT-compiled code on ARM CPUs warmly welcome!
